README.rst
.. image:: https://gitlab.com/ternaris/rosbags/badges/master/pipeline.svg
   :target: https://gitlab.com/ternaris/rosbags/-/commits/master
   :alt: pipeline status

.. image:: https://gitlab.com/ternaris/rosbags/badges/master/coverage.svg
   :target: https://gitlab.com/ternaris/rosbags/-/commits/master
   :alt: coverage report

.. image:: https://img.shields.io/pypi/pyversions/rosbags
   :alt: python versions

=======
Rosbags
=======

Rosbags is the **pure python** library for everything rosbag. It contains:

- **highlevel** easy-to-use interfaces,
- **rosbag2** reader and writer,
- **rosbag1** reader and writer,
- **extensible** type system with serializers and deserializers,
- **efficient converter** between rosbag1 and rosbag2,
- and more.

Rosbags does not have any dependencies on the ROS software stacks and can be used on its own or alongside ROS1 or ROS2.

Rosbags was developed for `MARV <https://gitlab.com/ternaris/marv-robotics>`_, which requires a fast, correct, and flexible library to read, manipulate, and write the various rosbag file formats.


Getting started
===============

Rosbags is published on PyPI and does not have any special dependencies. Simply install with pip::

   pip install rosbags


Read and deserialize messages from rosbag1 or rosbag2 files:

.. code-block:: python

   from pathlib import Path

   from rosbags.highlevel import AnyReader

   # create reader instance and open for reading
   with AnyReader([Path('/home/ros/rosbag_2020_03_24')]) as reader:
       connections = [x for x in reader.connections if x.topic == '/imu_raw/Imu']
       for connection, timestamp, rawdata in reader.messages(connections=connections):
            msg = reader.deserialize(rawdata, connection.msgtype)
            print(msg.header.frame_id)


Convert between rosbag versions::

   # Convert "foo.bag", result will be "foo/"
   rosbags-convert foo.bag

   # Convert "bar", result will be "bar.bag"
   rosbags-convert bar

   # Convert "foo.bag", save the result as "bar"
   rosbags-convert foo.bag --dst /path/to/bar

   # Convert "bar", save the result as "foo.bag"
   rosbags-convert bar --dst /path/to/foo.bag


Documentation
=============

Read the `documentation <https://ternaris.gitlab.io/rosbags/>`_ for further information.

.. end documentation


Contributing
============

Thank you for considering to contribute to rosbags.

To submit issues or create merge requests please follow the instructions provided in the `contribution guide <https://gitlab.com/ternaris/rosbags/-/blob/master/CONTRIBUTING.rst>`_.

By contributing to rosbags you accept and agree to the terms and conditions laid out in there.


Development
===========

Clone the repository and setup your local checkout::

   git clone https://gitlab.com/ternaris/rosbags.git
   
   cd rosbags
   python -m venv venv
   . venv/bin/activate
   
   pip install -r requirements-dev.txt
   pip install -e .


This creates a new virtual environment with the necessary python dependencies and installs rosbags in editable mode. The rosbags code base uses pytest as its test runner, run the test suite by simply invoking::

   pytest


To build the documentation from its source run sphinx-build::

   sphinx-build -a docs public


The entry point to the local documentation build should be available under ``public/index.html``.


Support
=======

Professional support is available from `Ternaris <https://ternaris.com>`_.


CHANGES.rst
.. _changes:

Changes
=======

0.9.16 - 2023-08-11
-------------------

- Support rosbag2 up to version 8 `#59`_
- Implement RIHS01 hashing of types
- Fix chunk size bug in rosbag1 writer `#58`_
- Fix handling of empty messages `#56`_
- Allow empty bags, chunks, and connections
- Avoid field name collisions with python keywords `#51`_
- Improve msg/idl type matching
- Improve some examples

.. _#51: https://gitlab.com/ternaris/rosbags/issues/51
.. _#56: https://gitlab.com/ternaris/rosbags/issues/56
.. _#58: https://gitlab.com/ternaris/rosbags/issues/58
.. _#59: https://gitlab.com/ternaris/rosbags/issues/59


0.9.15 - 2023-03-02
-------------------
- Refactor rosbag2 Reader for multipe storage backends
- Improve parsing of IDL files
- Handle bags contaning only connection records
- Add AnyReader to documentation
- Add initial MCAP reader for rosbag2 `#33`_

.. _#33: https://gitlab.com/ternaris/rosbags/issues/33


0.9.14 - 2023-01-12
-------------------
- Fix reader example in README `#40`_
- Flush decompressed files rosbag2.Reader
- Advertise Python 3.11 compatibility

.. _#40: https://gitlab.com/ternaris/rosbags/issues/40


0.9.13 - 2022-09-23
-------------------
- Fix parsing of comments in message definitions `#31`_
- Fix parsing of members starting with ``string`` in message definitions `#35`_
- Change lz4 compression level to 0 `#36`_
- Add include filters to rosbag conversion `#38`_
- Implement direct ros1 (de)serialization

.. _#31: https://gitlab.com/ternaris/rosbags/issues/31
.. _#35: https://gitlab.com/ternaris/rosbags/issues/35
.. _#36: https://gitlab.com/ternaris/rosbags/issues/36
.. _#38: https://gitlab.com/ternaris/rosbags/issues/38


0.9.12 - 2022-07-27
-------------------
- Add support for rosbag2 version 6 metadata `#30`_
- Enable rosbags-convert to exclude topics `#25`_

.. _#30: https://gitlab.com/ternaris/rosbags/issues/30
.. _#25: https://gitlab.com/ternaris/rosbags/issues/25


0.9.11 - 2022-05-17
-------------------
- Report start_time and end_time on empty bags


0.9.10 - 2022-05-04
-------------------
- Add support for multiple type stores
- Document which types are supported out of the box `#21`_
- Unify Connection and TopicInfo objects across rosbag1 and rosbag2
- Add experimental all-in-one reader for rosbag1, split rosbag1, and rosbag2
- Convert reader and writer .connection attribute from dict to list
- Add support for rosbag2 version 5 metadata `#18`_
- Speed up opening of rosbag1 files
- Fix serialization of empty message sequences `#23`_

.. _#18: https://gitlab.com/ternaris/rosbags/issues/18
.. _#21: https://gitlab.com/ternaris/rosbags/issues/21
.. _#23: https://gitlab.com/ternaris/rosbags/issues/23


0.9.9 - 2022-01-10
------------------
- Fix documentation code samples `#15`_
- Fix handling of padding after empty sequences `#14`_
- Support conversion from rosbag2 to rosbag1 `#11`_

.. _#11: https://gitlab.com/ternaris/rosbags/issues/11
.. _#14: https://gitlab.com/ternaris/rosbags/issues/14
.. _#15: https://gitlab.com/ternaris/rosbags/issues/15


0.9.8 - 2021-11-25
------------------
- Support bool and float constants in msg files


0.9.7 - 2021-11-09
------------------
- Fix parsing of const fields with string value `#9`_
- Parse empty msg definitions
- Make packages PEP561 compliant
- Parse msg bounded fields and default values `#12`_

.. _#9: https://gitlab.com/ternaris/rosbags/issues/9
.. _#12: https://gitlab.com/ternaris/rosbags/issues/12

0.9.6 - 2021-10-04
------------------
- Do not match msg separator as constant value


0.9.5 - 2021-10-04
------------------
- Add string constant support to msg parser


0.9.4 - 2021-09-15
------------------
- Make reader1 API match reader2
- Fix connection mapping for reader2 messages `#1`_, `#8`_

.. _#1: https://gitlab.com/ternaris/rosbags/issues/1
.. _#8: https://gitlab.com/ternaris/rosbags/issues/8

0.9.3 - 2021-08-06
------------------

- Add const fields to type classes
- Add CDR to ROS1 bytestream conversion
- Add ROS1 message definiton generator
- Use connection oriented APIs in readers and writers
- Add rosbag1 writer


0.9.2 - 2021-07-08
------------------

- Support relative type references in msg files


0.9.1 - 2021-07-05
------------------

- Use half-open intervals for time ranges
- Create appropriate QoS profiles for latched topics in converted bags
- Fix return value tuple order of messages() in documentation `#2`_
- Add type hints to message classes
- Remove non-default ROS2 message types
- Support multi-line comments in idl files
- Fix parsing of msg files on non-POSIX platforms `#4`_

.. _#2: https://gitlab.com/ternaris/rosbags/issues/2
.. _#4: https://gitlab.com/ternaris/rosbags/issues/4


0.9.0 - 2021-05-16
------------------

- Initial Release


src/rosbags/typesys/base.py
# Copyright 2020-2023  Ternaris.
# SPDX-License-Identifier: Apache-2.0
"""Types and helpers used by message definition converters."""

from __future__ import annotations

import json
import keyword
from enum import IntEnum, auto
from hashlib import sha256
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from typing import Any, Dict, List, Literal, Tuple, Union

    from .peg import Visitor
    from .register import Typestore

    Basetype = Union[str, Tuple[Literal['string'], int]]
    Constdefs = List[Tuple[str, str, Any]]
    Fielddesc = Union[Tuple[Literal[1], Basetype], Tuple[Literal[2], str],
                      Tuple[Literal[3, 4], Tuple[Union[Tuple[Literal[1], Basetype],
                                                       Tuple[Literal[2], str]], int]]]
    Fielddefs = List[Tuple[str, Fielddesc]]
    Typesdict = Dict[str, Tuple[Constdefs, Fielddefs]]


class TypesysError(Exception):
    """Parser error."""


class Nodetype(IntEnum):
    """Parse tree node types.

    The first four match the Valtypes of final message definitions.
    """

    BASE = auto()
    NAME = auto()
    ARRAY = auto()
    SEQUENCE = auto()

    LITERAL_STRING = auto()
    LITERAL_NUMBER = auto()
    LITERAL_BOOLEAN = auto()
    LITERAL_CHAR = auto()

    MODULE = auto()
    CONST = auto()
    STRUCT = auto()
    SDECLARATOR = auto()
    ADECLARATOR = auto()
    ANNOTATION = auto()
    EXPRESSION_BINARY = auto()
    EXPRESSION_UNARY = auto()


def normalize_fieldname(name: str) -> str:
    """Normalize field name.

    Avoid collisions with Python keywords.

    Args:
        name: Field name.

    Returns:
        Normalized name.

    """
    if keyword.iskeyword(name):
        return f'{name}_'
    return name


def parse_message_definition(visitor: Visitor, text: str) -> Typesdict:
    """Parse message definition.

    Args:
        visitor: Visitor instance to use.
        text: Message definition.

    Returns:
        Parsetree of message.

    Raises:
        TypesysError: Message parsing failed.

    """
    try:
        rule = visitor.RULES['specification']
        pos = rule.skip_ws(text, 0)
        npos, trees = rule.parse(text, pos)
        assert npos == len(text), f'Could not parse: {text!r}'
        return visitor.visit(trees)  # type: ignore
    except Exception as err:
        raise TypesysError(f'Could not parse: {text!r}') from err


TIDMAP = {
    'int8': 2,
    'uint8': 3,
    'int16': 4,
    'uint16': 5,
    'int32': 6,
    'uint32': 7,
    'int64': 8,
    'uint64': 9,
    'float32': 10,
    'float64': 11,
    'float128': 12,
    'char': 13,
    # 'wchar': 14,
    'bool': 15,
    'octet': 16,
    'string': 17,
    # 'wstring': 18,
    # 'fixed_string': 19,
    # 'fixed_wstring': 20,
    'bounded_string': 21,
    # 'bounded_wstring': 22,
}


def hash_rihs01(typ: str, typestore: Typestore) -> str:
    """Hash message definition.

    Args:
        typ: Message type name.
        typestore: Message type store.

    Returns:
        Hash value.

    """

    def get_field(name: str, desc: Fielddesc) -> dict[str, Any]:
        increment = 0
        capacity = 0
        string_capacity = 0
        subtype = ''
        if desc[0] == 3:
            increment = 48
            capacity = desc[1][1]
            typ, rest = desc[1][0]
        elif desc[0] == 4:
            count = desc[1][1]
            if count:
                increment = 96
                capacity = count
            else:
                increment = 144
            typ, rest = desc[1][0]
        else:
            typ, rest = desc

        # assert isinstance(rest, (str, tuple))
        if typ == 2:
            tid = increment + 1
            assert isinstance(rest, str)
            subtype = rest
            get_struct(subtype)
        else:
            if isinstance(rest, tuple):
                assert isinstance(rest[0], str)
                if rest[1]:
                    string_capacity = rest[1]
                    tid = increment + TIDMAP['bounded_string']
                else:
                    tid = increment + TIDMAP['string']
            else:
                tid = increment + TIDMAP[rest]

        return {
            'name': name,
            'type': {
                'type_id': tid,
                'capacity': capacity,
                'string_capacity': string_capacity,
                'nested_type_name': subtype,
            },
        }

    struct_cache = {}

    def get_struct(typ: str) -> dict[str, Any]:
        if typ not in struct_cache:
            struct_cache[typ] = {
                'type_name': typ,
                'fields': [
                    get_field(x, y) for x, y in typestore.FIELDDEFS[typ][1] or
                    [('structure_needs_at_least_one_member', (1, 'uint8'))]
                ],
            }
        return struct_cache[typ]

    dct = {
        'type_description': get_struct(typ),
        'referenced_type_descriptions': [y for x, y in sorted(struct_cache.items()) if x != typ],
    }

    digest = sha256(json.dumps(dct).encode()).hexdigest()
    return f'RIHS01_{digest}'
